// @alicloud/dysmsapi20170525/dist/models/GetCardSmsDetailsResponseBodyCardSendDetailDtorecords.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetCardSmsDetailsResponseBodyCardSendDetailDTORecords = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
class GetCardSmsDetailsResponseBodyCardSendDetailDTORecords extends $dara.Model {
    static names() {
        return {
            errCode: 'ErrCode',
            outId: 'OutId',
            phoneNumber: 'PhoneNumber',
            receiveDate: 'ReceiveDate',
            receiveType: 'ReceiveType',
            renderDate: 'RenderDate',
            renderStatus: 'RenderStatus',
            sendDate: 'SendDate',
            sendStatus: 'SendStatus',
            smsContent: 'SmsContent',
            templateCode: 'TemplateCode',
        };
    }
    static types() {
        return {
            errCode: 'string',
            outId: 'string',
            phoneNumber: 'string',
            receiveDate: 'string',
            receiveType: 'string',
            renderDate: 'string',
            renderStatus: 'number',
            sendDate: 'string',
            sendStatus: 'number',
            smsContent: 'string',
            templateCode: 'string',
        };
    }
    validate() {
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.GetCardSmsDetailsResponseBodyCardSendDetailDTORecords = GetCardSmsDetailsResponseBodyCardSendDetailDTORecords;
