// @alicloud/dysmsapi20170525/dist/models/SubmitSmsQualificationShrinkRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubmitSmsQualificationShrinkRequest = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
class SubmitSmsQualificationShrinkRequest extends $dara.Model {
    static names() {
        return {
            adminIDCardExpDate: 'AdminIDCardExpDate',
            adminIDCardFrontFace: 'AdminIDCardFrontFace',
            adminIDCardNo: 'AdminIDCardNo',
            adminIDCardPic: 'AdminIDCardPic',
            adminIDCardType: 'AdminIDCardType',
            adminName: 'AdminName',
            adminPhoneNo: 'AdminPhoneNo',
            businessLicensePicsShrink: 'BusinessLicensePics',
            bussinessLicenseExpDate: 'BussinessLicenseExpDate',
            certifyCode: 'CertifyCode',
            companyName: 'CompanyName',
            companyType: 'CompanyType',
            legalPersonIDCardNo: 'LegalPersonIDCardNo',
            legalPersonIDCardType: 'LegalPersonIDCardType',
            legalPersonIdCardBackSide: 'LegalPersonIdCardBackSide',
            legalPersonIdCardEffTime: 'LegalPersonIdCardEffTime',
            legalPersonIdCardFrontSide: 'LegalPersonIdCardFrontSide',
            legalPersonName: 'LegalPersonName',
            organizationCode: 'OrganizationCode',
            otherFilesShrink: 'OtherFiles',
            ownerId: 'OwnerId',
            qualificationName: 'QualificationName',
            remark: 'Remark',
            resourceOwnerAccount: 'ResourceOwnerAccount',
            resourceOwnerId: 'ResourceOwnerId',
            useBySelf: 'UseBySelf',
            whetherShare: 'WhetherShare',
        };
    }
    static types() {
        return {
            adminIDCardExpDate: 'string',
            adminIDCardFrontFace: 'string',
            adminIDCardNo: 'string',
            adminIDCardPic: 'string',
            adminIDCardType: 'string',
            adminName: 'string',
            adminPhoneNo: 'string',
            businessLicensePicsShrink: 'string',
            bussinessLicenseExpDate: 'string',
            certifyCode: 'string',
            companyName: 'string',
            companyType: 'string',
            legalPersonIDCardNo: 'string',
            legalPersonIDCardType: 'string',
            legalPersonIdCardBackSide: 'string',
            legalPersonIdCardEffTime: 'string',
            legalPersonIdCardFrontSide: 'string',
            legalPersonName: 'string',
            organizationCode: 'string',
            otherFilesShrink: 'string',
            ownerId: 'number',
            qualificationName: 'string',
            remark: 'string',
            resourceOwnerAccount: 'string',
            resourceOwnerId: 'number',
            useBySelf: 'boolean',
            whetherShare: 'boolean',
        };
    }
    validate() {
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.SubmitSmsQualificationShrinkRequest = SubmitSmsQualificationShrinkRequest;
