// @alicloud/dysmsapi20170525/dist/models/TagResourcesRequest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TagResourcesRequest = void 0;
// This file is auto-generated, don't edit it
const $dara = __importStar(require("@darabonba/typescript"));
const TagResourcesRequestTag_1 = require("./TagResourcesRequestTag");
class TagResourcesRequest extends $dara.Model {
    static names() {
        return {
            ownerId: 'OwnerId',
            prodCode: 'ProdCode',
            regionId: 'RegionId',
            resourceId: 'ResourceId',
            resourceOwnerAccount: 'ResourceOwnerAccount',
            resourceOwnerId: 'ResourceOwnerId',
            resourceType: 'ResourceType',
            tag: 'Tag',
        };
    }
    static types() {
        return {
            ownerId: 'number',
            prodCode: 'string',
            regionId: 'string',
            resourceId: { 'type': 'array', 'itemType': 'string' },
            resourceOwnerAccount: 'string',
            resourceOwnerId: 'number',
            resourceType: 'string',
            tag: { 'type': 'array', 'itemType': TagResourcesRequestTag_1.TagResourcesRequestTag },
        };
    }
    validate() {
        if (Array.isArray(this.resourceId)) {
            $dara.Model.validateArray(this.resourceId);
        }
        if (Array.isArray(this.tag)) {
            $dara.Model.validateArray(this.tag);
        }
        super.validate();
    }
    constructor(map) {
        super(map);
    }
}
exports.TagResourcesRequest = TagResourcesRequest;
