// @darabonba/typescript/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.XML = exports.URL = exports.SSEEvent = exports.Stream = exports.FileFormStream = exports.Form = exports.File = exports.Date = void 0;
__exportStar(require("./core"), exports);
var date_1 = require("./date");
Object.defineProperty(exports, "Date", { enumerable: true, get: function () { return __importDefault(date_1).default; } });
__exportStar(require("./error"), exports);
var file_1 = require("./file");
Object.defineProperty(exports, "File", { enumerable: true, get: function () { return __importDefault(file_1).default; } });
var form_1 = require("./form");
Object.defineProperty(exports, "Form", { enumerable: true, get: function () { return __importDefault(form_1).default; } });
Object.defineProperty(exports, "FileFormStream", { enumerable: true, get: function () { return form_1.FileFormStream; } });
__exportStar(require("./func"), exports);
__exportStar(require("./retry"), exports);
var stream_1 = require("./stream");
Object.defineProperty(exports, "Stream", { enumerable: true, get: function () { return __importDefault(stream_1).default; } });
Object.defineProperty(exports, "SSEEvent", { enumerable: true, get: function () { return stream_1.SSEEvent; } });
var url_1 = require("./url");
Object.defineProperty(exports, "URL", { enumerable: true, get: function () { return __importDefault(url_1).default; } });
var xml_1 = require("./xml");
Object.defineProperty(exports, "XML", { enumerable: true, get: function () { return __importDefault(xml_1).default; } });
